jonason/Parser.cpp

```diff
@@ -0,0 +1,226 @@
+static Out parse_json_object(Iterator, Iterator);
+static std::pair<std::string, JSONValue*> parse_json_object_key_value(Iterator, Iterator);
+static JSONValue::KeyType parse_json_object_key(Iterator, Iterator);
+static Out parse_json_array(Iterator, Iterator);
+static Out parse_json_string(Iterator, Iterator);
+
+std::unique_ptr<JSONValue> parse(const std::string& string)
+{
+    if (string.empty()) { throw ParseError::unexpected_eof; }
+
+    std::vector<Token> tokens;
+    tokenize(string, tokens);
+
+    if (tokens.empty()) { throw ParseError::unexpected_eof; }
+
+    auto begin = tokens.begin();
+    auto end = tokens.end();
+    return std::unique_ptr<JSONValue>(parse_json_value(begin, end));
+}
+
+std::unique_ptr<JSONValue> parse(std::istream& istream)
+{
+    std::vector<Token> tokens;
+    tokenize(istream, tokens);
+
+    if (tokens.empty()) { throw ParseError::unexpected_eof; }
+
+    auto begin = tokens.begin();
+    auto end = tokens.end();
+    return std::unique_ptr<JSONValue>(parse_json_value(begin, end));
+}
+
+static void parse_json_token(Token::Type token, Iterator iterator)
+{
+    if (iterator->tag == token) {
+        iterator++;
+    } else {
+        throw ParseError::unexpected_token(*iterator);
+    }
+}
+
+Out parse_json_value(Iterator iterator, Iterator end)
+{
+    switch (iterator->tag) {
+        case Token::DOUBLE_QUOTE:
+            return parse_json_string(iterator, end);
+        case Token::LITERAL:
+            switch (*iterator->value) {
+                case 't':
+                    if (strcmp(iterator->value, "true") == 0) {
+                        iterator++;
+                        return new JSONValue(true);
+                    } else {
+                        throw ParseError::unexpected_token(*iterator);
+                    }
+                    break;
+                case 'f':
+                    if (strcmp(iterator->value, "false") == 0) {
+                        iterator++;
+                        return new JSONValue(false);
+                    } else {
+                        throw ParseError::unexpected_token(*iterator);
+                    }
+                    break;
+                case 'n':
+                    if(strcmp(iterator->value, "null") == 0) {
+                        iterator++;
+                        return new JSONValue();
+                    } else {
+                        throw ParseError::unexpected_token(*iterator);
+                    }
+                    break;
+                default:
+                    double number = strtod(iterator->value, nullptr); // TODO: Handle endptr and out-of-range errors
+                    iterator++;
+                    return new JSONValue(number);
+            }
+        case Token::OBJECT_OPEN:
+            return parse_json_object(iterator, end);
+        case Token::ARRAY_OPEN:
+            return parse_json_array(iterator, end);
+        default:
+            throw ParseError::unexpected_token(*iterator);
```

Comment:
What are good practices for exception-free error handling?

Reply:
It'll vary a lot by codebase. I've mostly worked in codebases with COM-style HRESULT error codes but I'm sure you could construct something that looks more like swift's Error, though it'll probably not be as practical as the `try` keyword offers.